File: hask/day03.hs
#!/usr/bin/env stack
{- stack
  script
  --resolver lts-12.26
  --package hspec
  --package hspec-core
  --package split
  --package containers
-}
{-# LANGUAGE NamedFieldPuns #-}
import           Specs                          ( specFromExamples
                                                , specItem
                                                )
import           Test.Hspec                     ( describe
                                                , hspec
                                                , shouldBe
                                                )
import           Data.List.Split                ( splitOn )
import           Data.List                      ( sort )
import           Data.Maybe                     ( catMaybes )
import           Data.Set                       ( Set , fromList)

input :: IO (String, String)
input = do
  f <- readFile "../day03input.txt"
  let x : y : _ = lines f
  return (x, y)

readInput :: String -> [PathSegment]
readInput l = readOne <$> splitOn "," l
 where
  readOne :: String -> PathSegment
  readOne (x : xs) = case x of
    'U' -> (DUp, dist)
    'D' -> (DDown, dist)
    'L' -> (DLeft, dist)
    'R' -> (DRight, dist)
    _   -> error $ "could not parse: " ++ show [x : xs]
    where dist = read xs
  readOne "" = error "could not parse empty string"

data Direction = DUp | DDown | DLeft | DRight deriving (Show, Eq)

type Distance = Int
type PathSegment = (Direction, Distance)
type Point = (Int, Int)

data LineSegment = Horizontal { hY :: Int, hXs :: (Int, Int)} |
    Vertical { vX :: Int, vYs :: (Int, Int)} deriving (Show, Eq)

linesFromPaths :: [PathSegment] -> [LineSegment]
linesFromPaths ps = lfp (0, 0) ps []
 where
  lfp :: (Int, Int) -> [PathSegment] -> [LineSegment] -> [LineSegment]
  lfp _      []                 res = reverse res
  lfp (x, y) ((dir, dist) : ps) res = case dir of
    DUp    -> lfp (x, y + dist) ps $ Vertical x (y, y + dist) : res
    DDown  -> lfp (x, y - dist) ps $ Vertical x (y, y - dist) : res
    DLeft  -> lfp (x - dist, y) ps $ Horizontal y (x, x - dist) : res
    DRight -> lfp (x + dist, y) ps $ Horizontal y (x, x + dist) : res

intersect' :: Point -> Point -> Point -> Maybe Point
intersect' (px, py) rx ry = if intersects then Just (px, py) else Nothing
 where
  range (x1, x2) =
    let xmin = min x1 x2
        xmax = max x1 x2
    in  [xmin .. xmax]
  xsOverlap  = px `elem` range rx
  ysOverlap  = py `elem` range ry
  intersects = xsOverlap && ysOverlap

intersect :: (LineSegment, LineSegment) -> Maybe Point
intersect (Horizontal { hY, hXs }, Vertical { vX, vYs }) =
  intersect' (vX, hY) hXs vYs
intersect (Vertical { vX, vYs }, Horizontal { hY, hXs }) =
  intersect' (vX, hY) hXs vYs
intersect _ = Nothing

intersections :: ([LineSegment], [LineSegment]) -> [Point]
intersections (ls1, ls2) =
  filter (/= (0, 0)) $ catMaybes $ intersect <$> combinations
  where combinations = [ (x, y) | x <- ls1, y <- ls2 ]

distanceFromOrigo :: Point -> Int
distanceFromOrigo (x, y) = abs x + abs y

mapBoth :: (a -> b) -> (a, a) -> (b, b)
mapBoth f (x, y) = (f x, f y)

solve :: (String, String) -> Int
solve sp = minimum (distanceFromOrigo <$> intersections lsp)
 where
  psp = mapBoth readInput sp
  lsp = mapBoth linesFromPaths psp

tests = do
  describe "readInput" $ specFromExamples
    [ ("R8,U5,L5,D3", [(DRight, 8), (DUp, 5), (DLeft, 5), (DDown, 3)])
    , ("U7,R6,D4,L4", [(DUp, 7), (DRight, 6), (DDown, 4), (DLeft, 4)])
    ]
    (\(input, expected) ->
      specItem (show input ++ " should be parsed as: " ++ show expected)
        $          readInput input
        `shouldBe` expected
    )
  describe "intersections" $ specFromExamples
    [("R8,U5,L5,D3", "U7,R6,D4,L4", [(3, 3), (6, 5)]),
    ( "R75,D30,R83,U83,L12,D49,R71,U7,L72"
      , "U62,R66,U55,R34,D71,R55,D58,R83"
      , [(146,46),(155,4),(155,11),(158,-12)]
      )]
    (\(first, second, expected) ->
      specItem
          (  show first
          ++ " with "
          ++ second
          ++ " should yield: "
          ++ show expected
          )
        $ let lsp    = mapBoth (linesFromPaths . readInput) (first, second)
              actual = intersections lsp
          in  fromList actual `shouldBe` fromList expected
    )
  describe "solve" $ specFromExamples
    [ ("R8,U5,L5,D3", "U7,R6,D4,L4", 6)
    , ( "R75,D30,R83,U83,L12,D49,R71,U7,L72"
      , "U62,R66,U55,R34,D71,R55,D58,R83"
      , 159
      )
    , ( "R98,U47,R26,D63,R33,U87,L62,D20,R33,U53,R51"
      , "U98,R91,D20,R16,D67,R40,U7,R15,U6,R7"
      , 135
      )
    ]
    (\(first, second, expected) ->
      specItem
          (  show first
          ++ " with "
          ++ second
          ++ " should yield: "
          ++ show expected
          )
        $          solve (first, second)
        `shouldBe` expected
    )


main :: IO ()
main = do
  hspec tests
  inp <- input
  print $ solve inp

File: rust/src/day03.rs
use super::common::read_day_data;

#[derive(Debug)]
pub enum Direction {
    Up,
    Down,
    Left,
    Right,
}

#[derive(Debug)]
pub struct PathElement {
    direction: Direction,
    distance: u32,
}

type Path = Vec<PathElement>;

mod parsing {
    use super::{Direction, Path, PathElement};

    #[derive(Debug)]
    enum Error {
        InvalidDirection(String),
        InvalidDistance(String),
    }

    fn distance(s: &str) -> Result<u32, Error> {
        let res = s.parse();
        match res {
            Ok(x) => Ok(x),
            Err(_) => Err(Error::InvalidDistance(String::from(s))),
        }
    }

    fn direction_and_next(s: &str) -> Result<(Direction, &str), Error> {
        match s.split_at(1) {
            ("R", dist) => Ok((Direction::Right, dist)),
            ("U", dist) => Ok((Direction::Up, dist)),
            ("D", dist) => Ok((Direction::Down, dist)),
            ("L", dist) => Ok((Direction::Left, dist)),
            _ => Err(Error::InvalidDirection(String::from(s))),
        }
    }

    fn path_element(s: &str) -> Result<PathElement, Error> {
        let (dir, next) = direction_and_next(s)?;
        let dist = distance(next)?;
        Ok(PathElement {
            direction: dir,
            distance: dist,
        })
    }

    pub fn line(s: &str) -> Path {
        s.split(",")
            .map(path_element)
            .map(|rd| rd.expect("a path element"))
            .collect()
    }
}

pub fn input() -> Vec<Path> {
    read_day_data(3).lines().map(parsing::line).collect()
}

#[derive(PartialEq, Eq, Hash, Copy, Clone)]
pub struct Point {
    x: i32,
    y: i32,
}

pub mod part_1 {
    use super::*;

    macro_rules! solution_tests {
        ($($name:ident: $value:expr,)*) => {
        $(
            #[test]
            fn $name() {
                let ((input1, input2), expected) = $value;
                let paths = vec![input1, input2].iter().map(|&x| parsing::line(x)).collect();
                let result = solution(&paths);
                assert_eq!(Ok(expected), result);
            }
        )*
        }
    }

    solution_tests! {
        baby: (("R8,U5,L5,D3", "U7,R6,D4,L4"), 6),
        ex1: (("R75,D30,R83,U83,L12,D49,R71,U7,L72","U62,R66,U55,R34,D71,R55,D58,R83"),159),
        ex2: (("R98,U47,R26,D63,R33,U87,L62,D20,R33,U53,R51","U98,R91,D20,R16,D67,R40,U7,R15,U6,R7"),135),
    }

    use std::collections::HashSet;

    fn points_on_elem(pelm: &PathElement, mut p: &mut Point) -> HashSet<Point> {
        let mut res: HashSet<Point> = HashSet::new();
        let operation: fn(&mut Point) -> () = match pelm.direction {
            Direction::Up => |i| i.y += 1,
            Direction::Down => |i| i.y -= 1,
            Direction::Left => |i| i.x -= 1,
            Direction::Right => |i| i.x += 1,
        };
        for _i in 1..=pelm.distance {
            operation(&mut p);
            res.insert(p.clone());
        }
        res
    }
    fn points_on_path(p: &Path) -> HashSet<Point> {
        let mut cur = Point { x: 0, y: 0 };
        let mut res: HashSet<Point> = HashSet::new();
        for pelm in p {
            let new_points = points_on_elem(&pelm, &mut cur);
            let combined: HashSet<_> = res.union(&new_points).map(|&x| x).collect();
            res = combined;
        }
        res
    }
    fn manhattan_distance(p: &Point) -> i32 {
        p.x.abs() + p.y.abs()
    }

    pub fn solution(input: &Vec<Path>) -> Result<i32, String> {
        if input.len() != 2 {
            return Err(String::from("Expected exactly two paths"));
        }
        let px = points_on_path(&input[0]);
        let py = points_on_path(&input[1]);

        Ok(px.intersection(&py).map(manhattan_distance).min().unwrap())
    }
}

pub mod part_2 {
    use super::*;

    macro_rules! solution_tests {
        ($($name:ident: $value:expr,)*) => {
        $(
            #[test]
            fn $name() {
                let ((input1, input2), expected) = $value;
                let paths = vec![input1, input2].iter().map(|&x| parsing::line(x)).collect();
                let result = solution(&paths);
                assert_eq!(Ok(expected), result);
            }
        )*
        }
    }

    solution_tests! {
        ex1: (("R75,D30,R83,U83,L12,D49,R71,U7,L72","U62,R66,U55,R34,D71,R55,D58,R83"),610),
        ex2: (("R98,U47,R26,D63,R33,U87,L62,D20,R33,U53,R51","U98,R91,D20,R16,D67,R40,U7,R15,U6,R7"),410),
    }

    use std::collections::HashMap;
    use std::collections::HashSet;

    type PointDistanceMap = HashMap<Point, u32>;

    fn points_on_elem(
        pelm: &PathElement,
        mut map: &mut PointDistanceMap,
        mut p: &mut Point,
        mut i: &mut u32,
    ) -> () {
        let operation: fn(&mut Point) -> () = match pelm.direction {
            Direction::Up => |i| i.y += 1,
            Direction::Down => |i| i.y -= 1,
            Direction::Left => |i| i.x -= 1,
            Direction::Right => |i| i.x += 1,
        };
        for _i in 1..=pelm.distance {
            operation(&mut p);
            *i += 1;
            if !map.contains_key(p) {
                map.insert(p.clone(), *i);
            }
        }
    }
    fn points_on_path(p: &Path) -> PointDistanceMap {
        let mut cur = Point { x: 0, y: 0 };
        let mut res = HashMap::new();
        let mut dist = 0;
        for pelm in p {
            let new_points = points_on_elem(&pelm, &mut res, &mut cur, &mut dist);
        }
        res
    }

    fn path_distance(mpx: &PointDistanceMap, mpy: &PointDistanceMap, p: &Point) -> u32 {
        let dx = mpx.get(p).expect("point should be in map");
        let dy = mpy.get(p).expect("point should be in map");
        dx + dy
    }

    pub fn solution(input: &Vec<Path>) -> Result<u32, String> {
        if input.len() != 2 {
            return Err(String::from("Expected exactly two paths"));
        }
        let mpx = points_on_path(&input[0]);
        let mpy = points_on_path(&input[1]);

        let px: HashSet<Point> = mpx.keys().map(|&p| p).collect();
        let py: HashSet<Point> = mpy.keys().map(|&p| p).collect();

        Ok(px
            .intersection(&py)
            .map(|&p| path_distance(&mpx, &mpy, &p))
            .min()
            .unwrap())
    }
}
